test/distributed/optim/test_zero_redundancy_optimizer.py

```diff
@@ -204,6 +204,47 @@ def test_zero_grad(self):
         self.assertFalse(m.weight.grad)
         self.assertFalse(m.bias.grad)
 
+    def test_constructor(self):
+        """Check the robustness of the ZeroRedundancyOptimizer constructor by
+        passing different values for `params`"""
+        self.dist_init(self.rank)
+
+        m = torch.nn.Linear(1, 1)
+        # (input, expected error)
+        inputs = [
+            ([], ValueError),                           # empty parameter list
+            (torch.randn(1), TypeError),                # non-iterable: `torch.Tensor`
+            (1.2, TypeError),                           # non-iterable: `float`
+            ([{"params": m.parameters()}], TypeError),  # iterable of dict
+            (list(m.parameters()) + [42], TypeError),   # iterable containing non-`torch.Tensor`
+            (m.parameters(), None),                     # `params` as a generator
+            (list(m.parameters()), None)                # `params` as a list
+        ]
+
+        for (input, error) in inputs:
+            if (error):
+                with self.assertRaises(error):
+                    ZeroRedundancyOptimizer(input, optimizer_class=SGD, lr=0.1)
+            else:
+                ZeroRedundancyOptimizer(input, optimizer_class=SGD, lr=0.1)
```

Comment:
With this logic, if any of the inputs raise an unexpected error, the test fails and the remaining inputs are not tested. This is okay, but can be improved with [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) which allows you to go through all inputs and then fail at the end if >= 1 of them fail. Same comment applies to`test_same_param_type()` below.

Reply:
Resolved: `MultiProcessTestCase` does not support subtests.